tensorflow/python/data/util/structure_test.py

```diff
@@ -49,41 +50,716 @@
 # sure they are not executed before the (eager- or graph-mode) test environment
 # has been set up.
 #
+
+def _test_flat_structure_combinations():
+  cases = [
+      (
+          "Tensor",
+          lambda: constant_op.constant(37.0),
+          tensor_spec.TensorSpec,
+          [dtypes.float32],
+          [[]]
+      ),
+      (
+          "TensorArray",
+          lambda: tensor_array_ops.TensorArray(
+              dtype=dtypes.float32, element_shape=(3,),
+              size=0),
+          tensor_array_ops.TensorArraySpec,
+          [dtypes.variant],
+          [[]]
+      ),
+      (
+          "SparseTensor",
+          lambda: sparse_tensor.SparseTensor(
+              indices=[[3, 4]], values=[-1],
+              dense_shape=[4, 5]),
+          sparse_tensor.SparseTensorSpec,
+          [dtypes.variant],
+          [None]
+      ),
+      (
+          "RaggedTensor",
+          lambda: ragged_factory_ops.constant(
+              [[1, 2], [], [4]]),
+          ragged_tensor.RaggedTensorSpec,
+          [dtypes.variant],
+          [None]
+      ),
+      (
+          "Nested_0",
+          lambda: (constant_op.constant(37.0),
+                   constant_op.constant([1, 2, 3])),
+          tuple,
+          [dtypes.float32, dtypes.int32],
+          [[], [3]]
+      ),
+      (
+          "Nested_1",
+          lambda: {
+              "a": constant_op.constant(37.0),
+              "b": constant_op.constant([1, 2, 3])
+          },
+          dict,
+          [dtypes.float32, dtypes.int32],
+          [[], [3]]
+      ),
+      (
+          "Nested_2",
+          lambda: {
+              "a": constant_op.constant(37.0),
+              "b": (sparse_tensor.SparseTensor(
+                  indices=[[0, 0]], values=[1],
+                  dense_shape=[1, 1]), sparse_tensor.SparseTensor(
+                      indices=[[3, 4]], values=[-1],
+                      dense_shape=[4, 5]))
+          },
+          dict,
+          [dtypes.float32, dtypes.variant, dtypes.variant],
+          [[], None, None]
+      ),
+  ]
+  def reduce_fn(x, y):
+    name, value_fn, expected_structure, expected_types, expected_shapes = y
+    return x + combinations.combine(
+        value_fn=combinations.NamedObject(
+            "value_fn.{}".format(name), value_fn
+        ),
+        expected_structure=combinations.NamedObject(
+            "expected_structure.{}".format(name), expected_structure
+        ),
+        expected_types=combinations.NamedObject(
+            "expected_types.{}".format(name), expected_types
+        ),
+        expected_shapes=combinations.NamedObject(
+            "expected_shapes.{}".format(name), expected_shapes
+        )
```

Comment:
you should either 1) use a lambda and wrap it in `NamedObject` or 2) do not wrap the value in `NamedObject`. Wrapping a value that is not callable in `NamedObject` creates the need to use `_obj`, which should be avoided.

This comment applies throughout this PR.

Reply:
@jsimsa considering the lines mentioned above, if I refactor it to something like this:
```python
def reduce_fn(x, y):
    name, value_fn, expected_structure, expected_types, expected_shapes = y
    return x + combinations.combine(
        value_fn=combinations.NamedObject(
            "value_fn.{}".format(name), value_fn
        ),
        expected_structure=expected_structure,
        expected_types=expected_types,
        expected_shapes=expected_shapes
    )
```
I get the following error:
```console
raise DuplicateTestNameError(test_class_name, new_name, original_name)
absl.testing.parameterized.DuplicateTestNameError: Duplicate parameterized test name in StructureTest: generated test name 
'testFlatStructure_test_mode_eager_tfapiversion_1_expectedshapes__expectedstructure_classdict_expectedtypes_dtypevariant_valuefn_valuefnNested2' (generated from 
'testFlatStructure_test_mode_eager_tfapiversion_1_expectedshapes__expectedstructure_classdict_expectedtypes_dtypevariant_valuefn_valuefnNested2') already exists.
Consider using named_parameters() to give your tests unique names and/or renaming the conflicting test method.
```

Reply:
Will use lambda's only for this test case as others seem to work fine.

Reply:
addressed in the commit.